golem-component-service-base/src/service/component.rs

```diff
@@ -304,6 +413,142 @@ impl ComponentServiceDefault {
+            self.upload_protected_component(&versioned_component_id, data)
+        )?;
+
+        let component = Component {
+            component_size,
+            metadata: component_metadata,
+            created_at: Utc::now(),
+            component_type,
+            component_name: component_name.clone(),
+            versioned_component_id: versioned_component_id.clone(),
+            namespace: namespace.clone(),
+            files: uploaded_files.clone(),
+        };
+
+        let record = component
+            .clone()
+            .try_into()
+            .map_err(|e| ComponentError::conversion_error("record", e))?;
+
+        let result = self.component_repo.create(&record).await;
+
+        if let Err(RepoError::UniqueViolation(_)) = result {
+            Err(ComponentError::AlreadyExists(component_id.clone()))?;
+        }
+
+        self.component_compilation
+            .enqueue_compilation(component_id, component.versioned_component_id.version)
+            .await;
+
+        Ok(component)
+    }
+
+    async fn update_unchecked<Namespace>(
+        &self,
+        component_id: &ComponentId,
+        data: Vec<u8>,
+        component_type: Option<ComponentType>,
+        files: Option<Vec<InitialComponentFile>>,
+        namespace: &Namespace,
+    ) -> Result<Component<Namespace>, ComponentError>
+    where
+        Namespace: Display + TryFrom<String> + Eq + Clone + Send + Sync,
+        <Namespace as TryFrom<String>>::Error: Display + Debug + Send + Sync + 'static,
+    {
+        let created_at = Utc::now();
+
+        let metadata = ComponentMetadata::analyse_component(&data)
+            .map_err(ComponentError::ComponentProcessingError)?;
+
+        info!(namespace = %namespace, "Uploaded component - exports {:?}", metadata.exports);
+
+        let constraints = self.component_repo.get_constraint(component_id).await?;
+
+        let new_type_registry = FunctionTypeRegistry::from_export_metadata(&metadata.exports);
+
+        if let Some(constraints) = constraints {
+            let conflicts =
+                Self::find_component_metadata_conflicts(&constraints, &new_type_registry);
+            if !conflicts.is_empty() {
+                return Err(ComponentError::ComponentConstraintConflictError(conflicts));
+            }
+        }
+
+        let next_component = self
+            .component_repo
+            .get_latest_version(&component_id.0)
+            .await?
+            .filter(|c| c.namespace == namespace.to_string())
+            .ok_or(ComponentError::UnknownComponentId(component_id.clone()))
+            .and_then(|c| {
+                c.try_into()
+                    .map_err(|e| ComponentError::conversion_error("record", e))
+            })
+            .map(Component::next_version)?;
+
+        // Fallback to the files from the previous version if no files are provided.
+        let files_to_use = files.unwrap_or_else(|| next_component.files.clone());
```

Comment:
Please remind me of the spec while I review the PR. Is it about merging with the previous files (if none of the new files with the ones of previous version) ?

Reply:
Yeah, when no files are uploaded the set from the previous versions should be used.
To make my live simpler everywhere else here I just copy the data over from the previous version here (it's only the file paths + permissions, so quite small. All the actual data is deduplicated and in the blob_store)